### ISSUE TITLE
No way to catch exception 

### ISSUE THREAD

**Issue by nikitinsm** [alorence/django-modern-rpc#13]
Accroding to https://github.com/alorence/django-modern-rpc/blob/master/modernrpc/views.py#L97
and https://github.com/alorence/django-modern-rpc/blob/master/modernrpc/handlers/jsonhandler.py#L176

Looks like there is no way to raise Exception to top level. So there is no way to log it with django logging module or sentry for instance. In settings.DEBUG=True mode exception must be raised and NOT 'swallowed' or 'silented' by inner try/exception block and returned as normal response output

**Comment by nikitinsm** [comment]
The only way it to implement your own handler, witch is not handy way I think.

**Comment by alorence** [comment]
Hi Sergey, and thank you for your feedback.

Currently, exceptions is the mechanism used to return a correct error response response to XML-RPC or JSON-RPC client. A RPCException can be raised from modernrpc code at any moment if a common error is detected (unable to parse incoming payload, call to an unknown method, invalid params passed to the method, etc.). In addition, any exception raised from a RPC Method (your code) is catched and converted to RPCInternalError. This is mandatory to respect both XML-RPC and JSON-RPC standards.

----

```
So there is no way to log it with django logging module
```
Actually, default logger is used to print a warning or an error when an exception is raised. You can configure the logging system to handle these logs. See http://django-modern-rpc.readthedocs.io/en/latest/advanced/tips_and_tricks.html#enable-logging for more information.

Of course these messages may lack some information because the stacktrace and the underlying exception object is not kept. In such case, we could create a new logger to specifically handle exceptions. Example:
```
exceptions_logger = get_modernrpc_logger('modernrpc.exceptions')
# [...]

            try:
                if not handler.can_handle():
                    continue

                logger.debug('Request will be handled by {}'.format(handler_cls.__name__))

                result = handler.process_request()

                return handler.result_success(result)

            except AuthenticationFailed as e:
                # Customize HttpResponse instance used when AuthenticationFailed was raised
                exceptions_logger.exception('Authentication failed', e)
                return handler.result_error(e, HttpResponseForbidden)

            except RPCException as e:
                logger.warning('RPC exception raised: {}'.format(str(e)))
                exceptions_logger.exception('Authentication failed', e)
                return handler.result_error(e)

            except Exception as e:
                logger.error('Exception raised from a RPC method: {}'.format(str(e)))
                exceptions_logger.exception('Authentication failed', e)
                return handler.result_error(RPCInternalError(str(e)))
```

In such case, you should be able to configure logging for "modernrpc.exceptions" and dump information in a specific log file, your default log file, a mail to admins, etc.

----

```
In settings.DEBUG=True mode exception must be raised and NOT 'swallowed' or 'silented' by inner try/exception block and returned as normal response output
```
If I understand correctly, you want to be able to see a console stacktrace on error when developping, instead of ensuring the client get a clean error response? Am I right? That makes sense to me, but I think this kind of behavior should not be related to `settings.DEBUG` value, but to another specific settings constant.

----

In addition, I think it would be interesting to send some signals from modernrpc code on some specific event (like catched exception). In such case, you would be able to connect some receiver to these signals and execute the code you want accordingly.

----

To summarize, I propose to:

1. Create a new logger dedicated to exceptions, so you can configure the logging system to handle them like you want for your project (useful to track errors in production)
2. Add a setting value to completely disable exception catching & error response (useful in development)
2. Create and send some signals on various events, so you can execute additional code (in your case, when exceptions are raised)

Do you think these changes could solve your issue ?

**Comment by alorence** [comment]
Hi Sergey,

Sorry for the late answer, I am a bit for a few weeks.

I updated the library to improve generated log messages. In addition, a new setting value has been added:
``settings.MODERNRPC_LOG_EXCEPTION`` (which is now True by default).

If you configure your Django project to capture logs from module ``modernrpc``, you should be able to get information about errors in your RPC methods.

I use this version for a few days on a real project, and it works fine. In my case, the logging system use ``django.utils.log.AdminEmailHandler`` to send me a mail when exception is raised.

I didn't have a chance to test with Sentry, since I do not use it for now. Maybe you can test on your side to check if Sentry can catch errors from RPC methods with this configuration ?

Here is an example of my LOGGING config:

```
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse',
        },
        'require_debug_true': {
            '()': 'django.utils.log.RequireDebugTrue',
        },
    },
    'formatters': {
        'default': {
            'format': "[%(asctime)s] %(levelname)s [%(filename)s:%(funcName)s:%(lineno)s] %(message)s",
            'datefmt': "%Y-%m-%d %H:%M:%S"
        },
    },
    'handlers': {
        'default': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': log_dir('project.log'),
            'maxBytes': 1024 * 1024 * 1,  # 1 MB
            'backupCount': 5,
            'formatter': 'default'
        },
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler'
        },
        'console_debug': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'default',
            'filters': ['require_debug_true'],
        }
    },
    'loggers': {
        'modernrpc': {
            'handlers': ['default', 'console_debug', 'mail_admins', ],
            'level': 'INFO',
        }
    }
}
```

**Comment by alorence** [comment]
Hi again Sergey,

I finally found a few minutes to try Sentry with Raven on a python project.

It is possible to track exceptions from RPC methods by configuring the right logger. This also works with Raven and Sentry. Here is an example of the configuration I used:

```
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'sentry': {
            'level': 'ERROR',
            'class': 'raven.handlers.logging.SentryHandler',
            'dsn': "https://xxx:zzz@sentry.io/123456789",
        },
    },
    'loggers': {
        'modernrpc': {
            'handlers': ['sentry'],
            'level': 'INFO',
            'propagate': True,
        },
    }
}
```

With this very basic configuration, I can track and report errors to Sentry, even with current version of django-modern-rpc (0.9.0). In addition, I checked with current development version (not released yet) and the new ``settings.MODERNRPC_LOG_EXCEPTION``. In this case, Sentry can access to the full exception stacktrace which is more useful.

I close this issue for now, since you have a sample config to redirect modernrpc exceptions to sentry. I will open another ticket to improve the documentation and explain how error management can be configured.

Best regards !

**Comment by luzik** [comment]
I got this

Warning: Unable to load module "api.rpc_methods" declared in settings.MODERNRPC_METHODS_MODULES. Please ensure it is available and doesn't contain any error

and **no trace** with logger configured to handle modernrpc.

**Comment by alorence** [comment]
@luzik  to avoid noise for other participants, can you please open a new issue as this one was closed 3 years ago ?
